stake-pool/program/tests/withdraw.rs

```diff
@@ -151,3 +196,736 @@ async fn test_stake_pool_withdraw() {
+        mut banks_client,
+        payer,
+        recent_blockhash,
+        mut stake_pool_accounts,
+        validator_stake_account,
+        deposit_info,
+        _,
+        lamports_to_withdraw,
+    ) = setup().await;
+
+    // Create stake account to withdraw to
+    let user_stake_recipient = Keypair::new();
+
+    let new_authority = Pubkey::new_unique();
+    stake_pool_accounts.withdraw_authority = Keypair::new().pubkey();
+
+    let transaction_error = stake_pool_accounts
+        .withdraw_stake(
+            &mut banks_client,
+            &payer,
+            &recent_blockhash,
+            &user_stake_recipient.pubkey(),
+            &deposit_info.user_pool_account,
+            &validator_stake_account.stake_account,
+            &new_authority,
+            lamports_to_withdraw,
+        )
+        .await
+        .err()
+        .unwrap();
+
+    match transaction_error {
+        TransportError::TransactionError(TransactionError::InstructionError(
+            _,
+            InstructionError::Custom(error_index),
+        )) => {
+            let program_error = error::StakePoolError::InvalidProgramAddress as u32;
+            assert_eq!(error_index, program_error);
+        }
+        _ => panic!("Wrong error occurs while try to withdraw with wrong withdraw authority"),
+    }
+}
+
+#[tokio::test]
+async fn test_stake_pool_withdraw_with_wrong_token_program_id() {
+    let (
+        mut banks_client,
+        payer,
+        recent_blockhash,
+        stake_pool_accounts,
+        validator_stake_account,
+        deposit_info,
+        _,
+        lamports_to_withdraw,
+    ) = setup().await;
+
+    // Create stake account to withdraw to
+    let user_stake_recipient = Keypair::new();
+
+    let new_authority = Pubkey::new_unique();
+    let wrong_token_program = Keypair::new();
+
+    let mut transaction = Transaction::new_with_payer(
+        &[instruction::withdraw(
+            &id(),
+            &stake_pool_accounts.stake_pool.pubkey(),
+            &stake_pool_accounts.validator_stake_list.pubkey(),
+            &stake_pool_accounts.withdraw_authority,
+            &validator_stake_account.stake_account,
+            &user_stake_recipient.pubkey(),
+            &new_authority,
+            &deposit_info.user_pool_account,
+            &stake_pool_accounts.pool_mint.pubkey(),
+            &wrong_token_program.pubkey(),
+            &stake::id(),
+            lamports_to_withdraw,
+        )
+        .unwrap()],
+        Some(&payer.pubkey()),
+    );
+    transaction.sign(&[&payer], recent_blockhash);
+    let transaction_error = banks_client
+        .process_transaction(transaction)
+        .await
+        .err()
+        .unwrap();
+
+    match transaction_error {
+        TransportError::TransactionError(TransactionError::InstructionError(
+            _,
+            InstructionError::Custom(error_index),
+        )) => {
+            let program_error = error::StakePoolError::InvalidProgramAddress as u32;
```

Comment:
This should trigger an `IncorrectProgramId` error -- it looks like the `new_authority` might be throwing this off.  What is it for in this case?

Reply:
Right, the problem is in the smart contract, it throws a wrong exception and the test checks for it. Nice catch, fixing this.